### DISCUSSION TITLE
solve_lyap_lrcf not converged after 500 ADI steps

### DISCUSSION THREAD

**Discussion by Chunhui92** [pymor/pymor#1366] [Q&A]
Hi,
I am a fresher in MOR field, and currently using PyMOR for MNA(Modified Nodal Analysis). I tried to run PyMOR with an open-source data case **MNA example 1 (MNA_1.mat)** of [SLICOT](http://slicot.org/20-site/126-benchmark-examples-for-model-reduction), the system matrices A, E of size **578x578** are sparse and ill-conditioned. 

When I ran the code:
```
import scipy.io as spio
data = spio.loadmat('./data/MNA_1.mat')
E, A, B = data['E'], data['A'], data['B']
C = np.transpose(B)

fom = LTIModel.from_matrices(A, B, C, E=E, solver_options={'lyap_lrcf': 'lradi'})
hsv = fom.hsv()
```

A warning was returned as: 'Prescribed relative residual tolerance was not achieved (3.301581e+26 > 1.000000e-10) after 500 ADI steps.' Although the hsv array was returned, it had a lenght of **4500**.

I am wondering if my code has some mistakes or there is a compatability problem of PyMOR with severely ill-conditioned system matrices?

Than you!

**Comment by drittelhacker** [comment]
dear @Chunhui92,
welcome to MOR and pyMOR. 

As far as I can see, the conditioning is only part of the problem, the fact that MNA_1 is not stable is the other. If you check the properties and plots in http://eprints.maths.manchester.ac.uk/1040/1/ChahlaouiV02a.pdf, you can clearly see that the eigenvalues of (E,A) are all over the place. Unfortunately, LRADI is built on the assumption that the system is asymptotically stable (i.e. eigenvalues only in the left half-plane) and convergence to a certain extent relies on the fact that the spectrum is well away from the imaginary axis. Both are violated for this model. The fact that you see 4500 HSV is easily explained. The Gramian of the system is approximated by a low rank factorization Z*Z^T and LRADI adds 9 (the number of columns in B) new vectors to the solution factor Z in each iteration. So you see the 9*500 here. 

The benchmark has more difficulties in stock. looking at the smallest absolute  eigenvalues, these seem to come in "almost pairs" of real eigenvalues with positive and negative signs, which actually means that the Lyapunov Equation for the Gramian is extra badly conditioned, since exact mirror images would mean zero eigenvalues for the Lyapunov Operator.

Available solver for the large-scale case would likely all struggle with these.properties

**Reply by pmli** [reply]
Additionally, the E matrix in this example is singular, which is currently not supported in pyMOR.

**Comment by drittelhacker** [comment]
Then again it is not singular enough to apply standard (implicit) index reduction techniques.

**Reply by Chunhui92** [reply]
> Then again it is not singular enough to apply standard (implicit) index reduction techniques.

Thank you very much, now I know where the error comes from. 
If I understand right, the balanced truncation method with LRADI is not applicable to the MNA_1 case since its Lyapunov eqn is extra badly conditioned and it is unstable. Further, the implicit index reduction technique may be not applicable to this problem neither. Then, is it possible to apply Krylov subspace method to handle this case? Or generally, how can I use PyMOR for such cases? (Sorry for my naive questions, I lack knowledge on MOR...)

**Comment by lbalicki** [comment]
Hi @Chunhui92,

the fact that the E matrix in your example is singular, makes it difficult to apply the majority of system-theoretic MOR methods currently available in pyMOR. In particular some Krylov subspace based methods (e.g. [IRKA](https://docs.pymor.org/2020.2.0/generated/pymor.reductors.html#pymor.reductors.h2.IRKAReductor) or [bitangential Hermite interpolation](https://docs.pymor.org/2020.2.0/generated/pymor.reductors.html#pymor.reductors.interpolation.LTIBHIReductor)) will not work out of the box. Some of the theoretical issues and necessary adaptations are discussed [here](https://arxiv.org/abs/1301.4524).

With other Krylov subspace based methods like Padé-approximation/moment matching (see e.g. [this](https://www.cambridge.org/core/services/aop-cambridge-core/content/view/S0962492902000120)) there should not be any theoretical issues. However, there is no reductor for this in pyMOR yet and I think it would require some additions similar to the ones in the [pymor.algorithms.krylov](https://docs.pymor.org/2020.2.0/generated/pymor.algorithms.html#module-pymor.algorithms.krylov) subpackage.

On the pyMOR main branch (and probably in the next release) there is a [modal truncation reductor](https://github.com/pymor/pymor/blob/main/src/pymor/reductors/mt.py). This method will work for your example, but will not produce nice global error (bounds) like e.g. IRKA or balanced truncation.

**Reply by Chunhui92** [reply]
Thank you very much! I tried Krylov subspace method on the case, it worked very well.

On the other hand, I read from one [paper](https://link.springer.com/chapter/10.1007/3-540-27909-1_3) (see Remarks 3.3) that generalized balanced truncation can be applied to unstable system, by dividing the descriptor system into two subsystems for stable part and unstable part separately, then the stable part is reduced using balanced truncation while the unstable part is kept unmodified.
